program/src/processor.rs

```diff
@@ -644,3 +664,214 @@ pub fn process(program_id: &Pubkey, accounts: &[AccountInfo], input: &[u8]) -> P
+    // by the `calculate_rewards_per_token` function (tested above).
+    //
+    // The return type of `calculate_eligible_rewards` is limited to
+    // `u64::MAX`, but in order to determine the function's upper bounds for
+    // each input parameter, we must consider the maximum marginal reward per token
+    // (`current_accumulated_rewards_per_token`
+    //             - `last_accumulated_rewards_per_token`)
+    // and token account balance that this function can support.
+    //
+    // Since the marginal reward per token is at its maximum anytime a holder
+    // has a "last seen rate" (`last_accumulated_rewards_per_token`) of zero,
+    // we can evaluate in terms of `current_accumulated_rewards_per_token`,
+    // assuming the "last seen rate" to be zero. We will stick to this
+    // assumption in all references to `marginal_rewards_per_token` below.
+    //
+    // On its face, the maximum marginal reward per token is bound by
+    // `u128::MAX` - since both `current_accumulated_rewards_per_token` and
+    // `last_accumulated_rewards_per_token` are represented as `u128` integers.
+    // However, since the return value is capped at `u64::MAX`, we can perform
+    // the following arithmetic.
+    //
+    // Consider the original function:
+    //
+    //     eligible_rewards = (marginal_rewards_per_token * balance) / 1e9
+    //
+    // We can plug in `u64::MAX` for both `eligible_rewards` and `balance`
+    // to calculate the input `marginal_rewards_per_token` upper bound.
+    //
+    //     u64::MAX = (marginal_rewards_per_token * u64::MAX) / 1e9
+    //
+    // And evaluate to:
+    //
+    //     marginal_rewards_per_token = 1e9
+    //
+    // This means `calculate_eligible_rewards` can handle a maximum marginal
+    // reward per token of 1e9, or 1 lamport per token.
+    //
+    // But what does this mean as a constraint on the system as a whole? In
+    // other words, if a holder had 100% of the token supply and their last
+    // seen rate was zero, what's the maximum number of rewards the entire
+    // system can accumulate (in lamports) before this function would break?
+    //
+    // We can compute this value from the formula for
+    // `calculate_rewards_per_token`, which is represented below.
+    //
+    //     rewards_per_token = (reward * 1e9) / mint_supply
+    //
+    // Plugging in the values for maximum marginal reward per token and
+    // `u64::MAX` for token supply...
+    //
+    //     1e9 = (reward * 1e9) / u64::MAX
+    //
+    // ... we get:
+    //
+    //     reward = u64::MAX
+    //
+    // This means that the maximum lamports that can be paid into the system
+    // when a holder has 100% of the token supply and has never claimed is
+    // u64::MAX.
+    //
+    // This also means that the two functions - `calculate_rewards_per_token`
+    // and `calculate_eligible_rewards` share the same upper bound, since
+    // `calculate_rewards_per_token` expects a `u64` for rewards.
```

Comment:
I'm not sure I understand this -- what's the same upper bound that they share?

Reply:
They both would break past `u64::MAX` for accumulated system-level reward since you couldn't calculate rewards per token with that value either.

Reply:
Ah gotcha, makes sense!